### PR TITLE
Cache interceptor

### PR DESCRIPTION
**Pull request by sav007**
Part of #195 

This is the first version of http cache implementation.
This PR makes each call made with `ApolloClient` that has set `HttpCache` to be cached.

Cache control (or cache policy) with invalidation of cache will be implemented in next PR.

### CODE REVIEW

File: apollo-runtime/src/main/java/com/apollographql/android/HttpCacheInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.apollographql.android;
+
+import java.io.IOException;
+
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+final class HttpCacheInterceptor implements Interceptor {
+  static final String CACHE_CONTROL_HEADER = "APOLLO-CACHE-CONTROL";
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/HttpCacheInterceptor.java]: We need specs for this

File: apollo-runtime/src/main/java/com/apollographql/android/RealApolloCall.java [hunk 2]
```diff
@@ -126,6 +126,7 @@ private Request httpRequest(Operation operation) throws IOException {
         .post(requestBody)
         .header("Accept", ACCEPT_TYPE)
         .header("Content-Type", CONTENT_TYPE)
+        .header(HttpCacheInterceptor.CACHE_CONTROL_HEADER, "cache")
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/RealApolloCall.java]: for now all request are cached, we need specs for our custom cache control header

**Comment by brianPlummer** [comment]
looks good so far

File: apollo-runtime/src/main/java/com/apollographql/android/HttpCacheInterceptor.java (hunk 1 above)
Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/HttpCacheInterceptor.java]: One option: we can return a `CachingApolloClient` with one of:
```
client.newNetworkCall(operation)
client.newCall(operation,FORCE_NETWORK)
```

@martijnwalraven how does iOS bust through cache?

File: apollo-runtime/src/test/java/com/apollographql/android/IntegrationTest.java [hunk 3]
```diff
@@ -186,24 +211,42 @@
     assertThat(data.shop().products().edges().get(0).node().unsupportedCustomScalarTypeBool()).isEqualTo(Boolean.TRUE);
     assertThat(data.shop().products().edges().get(0).node().unsupportedCustomScalarTypeString()).isInstanceOf(String.class);
     assertThat(data.shop().products().edges().get(0).node().unsupportedCustomScalarTypeString()).isEqualTo("something");
+
+    okhttp3.Response cachedResponse = httpCache.read(((RealApolloCall) call).httpCall.request());
+    assertThat(cachedResponse).isNotNull();
+    assertThat(cachedResponse.body().source().readString(Charsets.UTF_8))
```
Comment (digitalbuddha) [inline apollo-runtime/src/test/java/com/apollographql/android/IntegrationTest.java]: `.readUtf8`

**Comment by digitalbuddha** [comment]
I noticed that you pulled the `CacheProxy` logic back into the `HTTPCache`, just an fyi thats how Square originally had it and then they moved it out.  I'm not crazy of it being in the Cache as it kind of breaks the single responsibility principle. Everything else looks good enough to continue

**Comment by digitalbuddha** [comment]
Does it make sense to slap a license on top of `httpcache` since it did start as an okhttp class?

**Comment by sav007** [comment]
`HttpCache` is not from okHttp, but `HeadersEntry` is copied, so I will add to it

File: apollo-runtime/src/main/java/com/apollographql/android/HttpCache.java [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+
+public final class HttpCache {
+  private static final int VERSION = 99991;
+  private static final int ENTRY_METADATA = 0;
+  private static final int ENTRY_BODY = 1;
+  private static final int ENTRY_COUNT = 2;
+  private final DiskLruCache cache;
+
+  public HttpCache(File directory, long maxSize, FileSystem fileSystem) {
+    this.cache = DiskLruCache.create(fileSystem, directory, VERSION, ENTRY_COUNT, maxSize);
```
Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/HttpCache.java]: I don't think an Lru cache is going to work for my use case. Can we have another constructor where I can pass in something the implements a cache or file system interface?

Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/HttpCache.java]: Sure will provide interface then

Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/HttpCache.java]: Could you have a public method to get the key again from a request. Thank you kindly.

File: apollo-runtime/src/main/java/com/apollographql/android/cache/ResponseHeaderRecord.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+import okhttp3.internal.platform.Platform;
+import okio.Buffer;
+import okio.BufferedSink;
+import okio.BufferedSource;
+import okio.ByteString;
+import okio.Okio;
+import okio.Source;
+
+/**
+ * Class was copied and modified from {@link okhttp3.Cache.Entry}
```
Comment (brianPlummer) [inline apollo-runtime/src/main/java/com/apollographql/android/cache/ResponseHeaderRecord.java]: thank you for this!! :)

**Pull request merged**